pkg/services/ngalert/accesscontrol/silences.go

```diff
@@ -0,0 +1,252 @@
+			ac.EvalAny(
+				ac.EvalPermission(instancesWrite),
+				ac.EvalPermission(silenceWrite, dashboards.ScopeFoldersProvider.GetResourceScopeUID(uid)),
+			),
+			readRuleSilenceEvaluator(uid),
+		)
+	}
+)
+
+type Silence interface {
+	GetRuleUID() *string
+}
+
+type RuleUIDToNamespaceStore interface {
+	GetNamespacesByRuleUID(ctx context.Context, orgID int64, uids ...string) (map[string]string, error)
+}
+
+type SilenceService struct {
+	genericService
+	store RuleUIDToNamespaceStore
+}
+
+func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *SilenceService {
+	return &SilenceService{
+		genericService: genericService{
+			ac: ac,
+		},
+		store: store,
+	}
+}
+
+// FilterByAccess filters the given list of silences based on the access control permissions of the user.
+// Global silence (one that is not attached to a particular rule) is considered available to all users.
+// For silences that are not attached to a rule, are checked against authorization.
+// This method is more preferred when many silences need to be checked.
+func (s SilenceService) FilterByAccess(ctx context.Context, user identity.Requester, silences ...Silence) ([]Silence, error) {
+	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
+	if err != nil || canAll { // return early if user can either read all silences or there is an error
+		return silences, err
+	}
+	canSome, err := s.HasAccess(ctx, user, readSomeSilenceEvaluator)
+	if err != nil || !canSome {
+		return nil, err
+	}
+	result := make([]Silence, 0, len(silences))
+	keys := make(map[string][]Silence, len(silences))
+	for _, silence := range silences {
+		ruleUID := silence.GetRuleUID()
+		if ruleUID == nil { // if this is a global silence
+			result = append(result, silence)
+			continue
+		}
+		key := *ruleUID
+		keys[key] = append(keys[key], silence)
+	}
+	if len(keys) == 0 { // if only global silences are provided no need in other checks
+		return result, nil
+	}
+	namespaces, err := s.store.GetNamespacesByRuleUID(ctx, user.GetOrgID(), maps.Keys(keys)...)
+	if err != nil {
+		return nil, err
+	}
+
+	for key, silence := range keys {
+		ns, ok := namespaces[key]
+		if !ok {
+			continue
+		}
+		hasAccess, err := s.HasAccess(ctx, user, readRuleSilenceEvaluator(ns))
+		if err != nil {
+			return nil, err
+		}
+		if !hasAccess {
+			continue
+		}
+		result = append(result, silence...)
+	}
+	return result, nil
+}
+
+// AuthorizeReadSilence checks if user has access to read a silence
+func (s SilenceService) AuthorizeReadSilence(ctx context.Context, user identity.Requester, silence Silence) error {
+	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
+	if canAll || err != nil { // return early if user can either read all silences or there is error
+		return err
+	}
+
+	can, err := s.HasAccess(ctx, user, readSomeSilenceEvaluator)
+	if err != nil {
+		return err
+	}
+	if !can { // if user cannot read global silences, then it does not have silence permissions at all.
```

Comment:
NIT: This comment doesn't seem accurate. Maybe just:
```suggestion
	if !can { // User does not have silence permissions at all.
```